README
Author: Bas Brugman

Company: http://www.visionnaire.nl

Weathervision is a  weather forecast script for Conky written in Ruby.
The weather data is extracted from http://www.wunderground.com/ for my personal use.


TODO........
text_buffer_size conky.......... 4096 make sure



CHANGELOG.md
02-05-2013
==========

Version 2.0
-------------
- Rewrote the parsing of weather data: now using the wunderground.com JSON free developer API. 
- Added another template for a horizontal conky layout
- Removed the text-based (.otf fonts) templates and functionality

25-07-2011
==========

Version 1.0
-------------
- Initial release


spec/spec_helper.rb
path = File.expand_path(File.dirname(__FILE__) + "/../lib/")
$LOAD_PATH.unshift(path) unless $LOAD_PATH.include?(path)
Bundler.require(:default, :development, :test)
# requires supporting files with custom matchers and macros, etc, in ./support/ and its subdirectories.
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}
RSpec.configure do |config|
  #suppress all output to only show relevant test data
  config.before(:all) do
    $stdout = StringIO.new
  end
end


spec/unit/forecast_parser_spec.rb
require 'spec_helper'

module Weathervision
  describe ForecastParser do
    let(:test_options_image) { { "apikey" => ENV["apikey"], "mode" => "himage", "current_location" => "vlist", "forecast_location" => "vlist", "radar_location" => "http://www.knmi.nl/neerslagradar/images/meest_recente_radarloop451.gif"} }
    let(:test_options_image_without_radar) { { "apikey" => ENV["apikey"], "mode" => "himage", "current_location" => "vlist", "forecast_location" => "vlist", "radar_location" => ""} }

    it "should have a template which outputs paths to images" do
      parser = ForecastParser.new(test_options_image)
      parser.parse
      parser.conky_output.should include(".png")
    end

    it "should call the right image methods according the options hash" do
      parser = ForecastParser.new(test_options_image)
      parser.should_receive(:calc_weather_icon).once
      parser.should_receive(:calc_wind_icon).once
      parser.parse
    end

    it "should process a radar image if set" do
      parser = ForecastParser.new(test_options_image)
      parser.should_receive(:parse_radar).once
      parser.parse
    end

    it "should not process a radar image when disabled" do
      parser = ForecastParser.new(test_options_image_without_radar)
      parser.should_not_receive(:parse_radar)
      parser.parse
    end

    it "should return the right color symbols for the wind speed" do
      parser = ForecastParser.new(test_options_image)
      parser.get_color("100").should == :red
      parser.get_color(0.2).should == :empty
      parser.get_color("1024").should == :empty
      parser.get_color(20.234).should == :yellow
      parser.get_color("13.5").should == :green
    end

    it "should merge the right wind icon name" do
      parser = ForecastParser.new(test_options_image)  
      parser.instance_eval { @forecast = { "0" => { "windspeed" => "1", "winddir" => "East" } } }
      parser.instance_eval { @current = { "windspeed" => "30", "winddir" => "NW" } }
      parser.calc_wind_icon
      parser.current["wind_icon"].should =~ /51\.png/
      parser.forecast["0"]["wind_icon"].should =~ /09\.png/
    end

    it "should merge the right weather icon name" do
      parser = ForecastParser.new(test_options_image)  
      parser.instance_eval { @forecast = { "0" => { "conditions" => "Fog" } } }
      parser.calc_weather_icon
      parser.forecast["0"]["weather_icon"].should =~ /20\.png/
    end

    it "should take a default icon when unknown weather conditions" do
      parser = ForecastParser.new(test_options_image)  
      parser.instance_eval { @forecast = { "0" => { "conditions" => "Change Rain" } } }
      parser.calc_weather_icon
      parser.forecast["0"]["weather_icon"].should =~ /25\.png/
    end

  end
end


spec/integration/forecast_parser_spec.rb
require 'spec_helper'

module Weathervision
  describe ForecastParser do
    let(:test_options_image) { { "apikey" => ENV["apikey"], "mode" => "himage", "current_location" => "vlist", "forecast_location" => "vlist", "radar_location" => "http://www.knmi.nl/neerslagradar/images/meest_recente_radarloop451.gif"} }

    it "should parse a weather forecast feed over the internet" do
      doc = HTTParty.get("http://api.wunderground.com/api/#{test_options_image['apikey']}/conditions/lang:EN/q/#{test_options_image['current_location']}.json")
      doc.class.should ==  HTTParty::Response
    end

    it "should parse a weather current feed over the internet" do
      doc = HTTParty.get("http://api.wunderground.com/api/#{test_options_image['apikey']}/forecast10day/lang:EN/q/#{test_options_image['forecast_location']}.json")
      doc.class.should == HTTParty::Response
    end

    it "should download and manipulate a radar image" do
      parser = ForecastParser.new(test_options_image)
      parser.parse
      File.size?(RADAR_PATH + '/radar.gif').should_not be_nil
    end

    it "should cleanup temporary images" do
      parser = ForecastParser.new(test_options_image)
      parser.parse
      Dir.glob("#{RADAR_PATH}/*.gif").size.should == 1
    end

    it "should parse the forecast weather conditions" do
      parser = ForecastParser.new(test_options_image)
      parser.forecast_json = mock(HTTParty::Response).as_null_object
      parser.parse_forecast
    end

    it "should parse the current weather conditions" do
      parser = ForecastParser.new(test_options_image)
      parser.parse_json
      parser.parse_current
      parser.current.keys.size.should == 8
    end

  end
end


lib/weathervision.rb
#external deps
require "json"
require "httparty"
require "thor"
require "erb"

#internal deps
require "weathervision/forecast_parser"
require "weathervision/cli"

require "pry"
require "debugger"
require "pry-debugger"

#wrapping namespace
module Weathervision
  # constant variable pointing to the absolute path where the (optional) satellite image will be saved 
  RADAR_PATH = File.expand_path("../assets/radars", File.dirname(__FILE__))
  # constant variable pointing to the absolute path where the weather icons are
  WEATHER_PATH = File.expand_path("../assets/images/weathericons", File.dirname(__FILE__))
  # constant variable pointing to the absolute path where the wind icons are
  BEARING_PATH = File.expand_path("../assets/images/bearingicons", File.dirname(__FILE__))
  # constant variable pointing to the absolute path where the image and text templates are
  TEMPLATE_PATH = File.expand_path("../templates", File.dirname(__FILE__))
end



lib/weathervision/cli.rb
module Weathervision
  class CLI < Thor

    desc "parse [options]", "Parse weather data" 
    method_option :mode, :type => :string, :required => true, :desc => "Mode: use 'vimage' or 'himage for vertical or horizontal layout'", :default => 'vimage'
    method_option :apikey, :type => :string, :required => true, :desc => "wunderground.com json API key"
    method_option :current_location, :type => :string, :required => true, :desc => "wunderground.com current weather location query string", :default => 'vlist'
    method_option :forecast_location, :type => :string, :required => true, :desc => "wunderground.com forecast weather location query string", :default => 'vlist'
    method_option :radar_location, :type => :string, :required => false, :desc => "URL of a radar image (Animated GIF)", :default => "http://www.knmi.nl/neerslagradar/images/meest_recente_radarloop451.gif"
    def parse
      parser = Weathervision::ForecastParser.new(options)
      parser.parse
    end
  end
end



lib/weathervision/forecast_parser.rb
module Weathervision #:nodoc:

  # The main weather parsing class
  class ForecastParser
    attr_reader :wind_icons, :image_icons, :forecast, :current, :conky_output, :radar
    attr_accessor :forecast_json, :current_json, :astronomy_json

    # Constructor method: predefines the data hashes and sets the user's options coming in
    # from Thor's cli api
    def initialize(options)
      @forecast, @current, @radar, @options = {}, {}, nil, options
      @himage_tpl = "#{TEMPLATE_PATH}/weathervision_himage.erb"
      @vimage_tpl = "#{TEMPLATE_PATH}/weathervision_vimage.erb"
      @wind_icons = {
        "VAR" => { empty: "00", green: "01", yellow: "02", orange: "03", red: "04" },
        "North" => { empty: "00", green: "05", yellow: "21", orange: "37", red: "53" },
        "NNE" => { empty: "00", green: "06", yellow: "22", orange: "38", red: "54" },
        "NE" => { empty: "00", green: "07", yellow: "23", orange: "39", red: "55" },
        "ENE" => { empty: "00", green: "08", yellow: "24", orange: "40", red: "56" },
        "East" => { empty: "00", green: "09", yellow: "25", orange: "41", red: "57" },
        "ESE" => { empty: "00", green: "10", yellow: "26", orange: "42", red: "58" },
        "SE" => { empty: "00", green: "11", yellow: "27", orange: "43", red: "59" },
        "SSE" => { empty: "00", green: "12", yellow: "28", orange: "44", red: "60" },
        "South" => { empty: "00", green: "13", yellow: "29", orange: "45", red: "61" },
        "SSW" => { empty: "00", green: "14", yellow: "30", orange: "46", red: "62" },
        "SW" => { empty: "00", green: "15", yellow: "31", orange: "47", red: "63" },
        "WSW" => { empty: "00", green: "16", yellow: "32", orange: "48", red: "64" },
        "West" => { empty: "00", green: "17", yellow: "33", orange: "49", red: "65" },
        "WNW" => { empty: "00", green: "18", yellow: "34", orange: "50", red: "66" },
        "NW" => { empty: "00", green: "19", yellow: "35", orange: "51", red: "67" },
        "NNW" => { empty: "00", green: "20", yellow: "36", orange: "52", red: "68" }
      }
      @image_icons = {
        "Chance of Flurries" => "13",
        "Chance of Rain" => "11",
        "Chance of Freezing Rain" => "10",
        "Chance of Sleet" => "18",
        "Chance of Snow" => "14",
        "Chance of Thunderstorms" => "38",
        "Chance of a Thunderstorm" => "38",
        "Clear" => "32",
        "Cloudy" => "26",
        "Flurries" => "13",
        "Fog" => "20",
        "Haze" => "21",
        "Mostly Cloudy" => "28",
        "Mostly Sunny" => "34",
        "Partly Cloudy" => "34",
        "Partly Sunny" => "30",
        "Freezing Rain" => "10",
        "Rain" => "12",
        "Sleet" => "18",
        "Snow" => "16",
        "Sunny" => "32",
        "Thunderstorms" => "03",
        "Thunderstorm" => "03",
        "Unknown" => "25",
        "Overcast" => "26",
        "Scattered Clouds" => "30"
      }
    end

    # Main processing function which starts the actual parsing and displaying
    def parse
      parse_json && parse_forecast && parse_current && show_image_version
    end

    # Final method which parses and outputs the image template
    def show_image_version
      if @options["mode"] == "himage"
        tpl = @himage_tpl
      elsif @options["mode"] == "vimage"
        tpl = @vimage_tpl
      else 
        raise "invalid template name" + @options["mode"]
      end
      calc_weather_icon
      calc_wind_icon
      parse_radar unless @options["radar_location"].empty?
      erb = ERB.new(File.read(tpl), 0, '>')
      @conky_output = erb.result(binding)
      puts @conky_output
    end

    # Fetches the optional radar animated GIF
    # Imagemagick's convert tool is used to make it a cropped single image
    def parse_radar
      outfile = RADAR_PATH + "/radar_#{Time.now.to_i}_#{Time.now.strftime("%d%m%Y")}.gif"
      %x(wget #{@options["radar_location"]} -O #{outfile} 2>&1 >/dev/null)
      %x(convert #{outfile} -flatten -repage '0x0' -crop '259x357+89+0' +repage #{RADAR_PATH + '/radar.gif'})
      # clear temp files and assign latest radar image
      Dir.glob("#{RADAR_PATH}/*.gif") do |filename|
        if filename =~ /radar\.gif/
          @radar = filename
        else
          File.delete(filename)
        end
      end
    end

    # Fetches the right weather icon (a PNG image)
    def calc_weather_icon
      @forecast.keys.each do |period|
        @forecast[period].merge!("weather_icon" => WEATHER_PATH + "/" + (@image_icons[@forecast[period]["conditions"]] || @image_icons["Unknown"]) + ".png")
      end
    end

    # Fetches the right wind icon name based on the wind direction and wind speed and adds
    # it to the hashes
    def calc_wind_icon
      @forecast.keys.each do |period|
        if @forecast[period].keys.include?("windspeed")
          color = get_color @forecast[period]["windspeed"]
          @forecast[period].merge!("wind_icon" => BEARING_PATH + "/" + (@wind_icons[@forecast[period]["winddir"]][color] || @wind_icons["VAR"][color])  + ".png")
        end
      end
      color = get_color @current["windspeed"]
      @current.merge!("wind_icon" => BEARING_PATH + "/" + (@wind_icons[@current["winddir"]][color] || @wind_icons["VAR"][color]) + ".png")
    end

    # Calculates the color to use as the arrow in the wind compass.
    # The numbers represent km/h:
    # 
    # * 0 means no wind
    # * 1 till 19 means light/gentle breeze
    # * 20 till 25 means moderate breeze
    # * 26 till 39 means fresh breeze
    # * 40 till 256 means strong wind, gale or even storms
    def get_color windspeed
      color = case windspeed.to_i
        when 0 then :empty 
        when 1..19 then :green 
        when 20..25 then :yellow 
        when 26..39 then :orange
        when 40..256 then :red 
      else
       :empty 
      end
    end

    # Parses the forecast weather conditions json feed and fills @forecast with 
    # the upcoming weather conditions (including today)
    def parse_forecast
      @forecast_json["forecast"]["simpleforecast"]["forecastday"].each do |day|
        hsh = {}
        hsh['weekday'] = day["date"]["weekday_short"]
        hsh['high'] = day["high"]["celsius"]
        hsh['low'] = day["low"]["celsius"]
        hsh['conditions'] = day["conditions"]
        hsh['windspeed'] = day["maxwind"]["kph"]
        hsh['winddir'] = day["maxwind"]["dir"]
        @forecast[day["period"]] = hsh # periods 1 to 10
      end
    end

    # Parses the current weather condition json feed and fills @current with the data
    def parse_current
      attr = @current_json['current_observation']
      @current['city'] = attr['observation_location']['city']
      @current['temp'] = attr['temp_c'].to_s
      @current['windspeed'] = attr['wind_kph']
      @current['winddir'] = attr['wind_dir']
      @current['winddegrees'] = attr['wind_degrees']
      @current['humidity'] = attr['relative_humidity']
      moon = @astronomy_json['moon_phase']
      @current['sunrise'] = moon['sunrise']['hour'] +  ':' + moon['sunrise']['minute']
      @current['sunset'] = moon['sunset']['hour'] +  ':' + moon['sunset']['minute']
    end

    # Parses the actual json data and sets the instance variables
    def parse_json
      @current_json = HTTParty.get("http://api.wunderground.com/api/#{@options['apikey']}/conditions/lang:EN/q/#{@options['current_location']}.json")
      @astronomy_json = HTTParty.get("http://api.wunderground.com/api/#{@options['apikey']}/astronomy/lang:EN/q/#{@options['current_location']}.json")
      @forecast_json = HTTParty.get("http://api.wunderground.com/api/#{@options['apikey']}/forecast10day/lang:EN/q/#{@options['forecast_location']}.json")
    end

  end
end
